packages/expo-network/android/src/main/java/expo/modules/network/NetworkModule.kt

```diff
@@ -1,31 +1,80 @@
 package expo.modules.network
 
-import expo.modules.core.Promise
-import expo.modules.core.ExportedModule
-import expo.modules.core.interfaces.ExpoMethod
-import expo.modules.core.interfaces.RegistryLifecycleListener
-
-import android.util.Log
-import android.os.Build
-import android.os.Bundle
 import android.content.Context
+import android.net.ConnectivityManager
+import android.net.NetworkCapabilities
 import android.net.NetworkInfo
 import android.net.wifi.WifiInfo
 import android.net.wifi.WifiManager
-import android.net.ConnectivityManager
-import android.net.NetworkCapabilities
+import android.os.Build
+import android.os.Bundle
 import android.provider.Settings
-
-import java.lang.Exception
+import android.util.Log
+import expo.modules.kotlin.Promise
+import expo.modules.kotlin.exception.Exceptions
+import expo.modules.kotlin.modules.Module
+import expo.modules.kotlin.modules.ModuleDefinition
 import java.math.BigInteger
 import java.net.InetAddress
 import java.net.UnknownHostException
 import java.nio.ByteOrder
 
-private const val NAME = "ExpoNetwork"
 private val TAG = NetworkModule::class.java.simpleName
 
-class NetworkModule(private val appContext: Context) : ExportedModule(appContext), RegistryLifecycleListener {
+class NetworkModule : Module() {
+  private val context: Context
+    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
+
+  override fun definition() = ModuleDefinition {
+    Name("ExpoNetwork")
+
+    AsyncFunction("getNetworkStateAsync") { promise: Promise ->
+      val result = Bundle()
+      val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+      try {
+        if (Build.VERSION.SDK_INT < 29) { // use getActiveNetworkInfo before api level 29
+          val netInfo = connectivityManager.activeNetworkInfo
+          val connectionType = getConnectionType(netInfo)
+
+          result.apply {
+            putBoolean("isInternetReachable", netInfo!!.isConnected)
+            putString("type", connectionType.value)
+            putBoolean("isConnected", connectionType.isDefined)
+          }
+
+          promise.resolve(result)
+        } else {
+          val network = connectivityManager.activeNetwork
+          val isInternetReachable = network != null
+
+          val connectionType = if (isInternetReachable) {
+            val netCapabilities = connectivityManager.getNetworkCapabilities(network)
+            getConnectionType(netCapabilities)
+          } else {
+            null
+          }
+
+          result.apply {
+            putString("type", connectionType?.value ?: NetworkStateType.NONE.value)
+            putBoolean("isInternetReachable", isInternetReachable)
+            putBoolean("isConnected", connectionType != null && connectionType.isDefined)
+          }
+          promise.resolve(result)
+        }
+      } catch (e: Exception) {
+        throw NetworkAccessException()
```

Comment:
would be better to pass the exception into the `NetworkAccessException`. the [`CodedException` ](https://github.com/expo/expo/blob/77aca4d3c9a8c43d058c5e272a3be99d6ef20352/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/exception/CodedException.kt#L16) has an optional `cause` parameter for this.